src/lib/Transform.ts

```diff
@@ -229,11 +236,27 @@ export function batch<T>(
 				flushError = e;
 			}
 
-			await writable.end(error, readable.result().catch(noop));
-			throwIfThrowable(earlyFlushError);
-			throwIfThrowable(flushError);
+			const toThrow = earlyFlushError || flushError;
+
+			await writable.end(
+				error,
+				toThrow
+					? readable.result().catch((e) => {
```

Comment:
Is all of this still necessary after #52 ?

Reply:
Tried it out; in my tests, I have to end the source with `.end(undefined, Promise.resolve())` to stop an error in `flush` from appearing in the result of `.forEach()`.

Thought about that for a while. I want to make sure I have this right:

The purpose of `result()` is to synchronize the end-states of all parts of a stream. Now with #52, it also has the role of ensuring that errors passed to `end()` throw downstream, and errors thrown by `end()` throw upstream.

This mirrors the behaviour of `write()`. When `write()` throws an error, the caller can resolve the error in some way or throw it upstream (if there is an upstream to throw to). It can also "throw it back" by calling `end()` on the downstream with the error (which makes sense to do if we're at the source, i.e., no upstream).

The behaviour of `end()` is necessarily different, though. When the upstream gets an error by calling `end()`, it also can resolve it or possibly throw it upstream. What it can't do is "throw it back", because it's already called `end()` and cannot call it again.

With #52, a stream can "throw it back" from `end()` by not including a `result` for `end()` to wait for. This basically signals that we're at the source, so uncaught errors have to bubble back down. In contrast, by including `result`, we indicate that the final outcome will depend on the fate of the error from `end()` as it bubbles up toward the source.

It seems to me that there is a missing capacity here, which is: how to resolve an error from `end()` when we're already at the source? We can do this if it's a failed call to `write()`.

Using `Promise.resolve()` works, but it's clearly a hack—I'm just "lying" to the ender, telling it that there's an infinitely forgiving source still farther upstream that will take care of everything!

One could also do something like:
```
const sourceResult = defer();
await source.write(1);
try {
  await source.end(undefined, sourceResult.promise);
  sourceResult.resolve();
} catch (e) {
  if (e.message === "I'm just in a bad mood") {
     // handle
     sourceResult.resolve()
  } else {
     sourceResult.reject()
  }
}
```

This is more "honest" in terms of the idiom, but not very easy to follow. (Maybe we could have a library function that wraps this behaviour? `endSource(error: Error | undefined, preemptEnderError: (enderError: Error) => void`?)

Back to the behaviour of batch, specifically: my conclusion is that things are working as they should. In some ways it seems like the `ender` of `batch` should be handling errors from `flush` outside of `ender`, since they're really `write()`-related, not `end()`-related. But that is a mixing of concerns that we've already committed to. My concerns here are really about #52 and not about `batch()`.

Reply:
TL;DR I think you generally understood correctly. I'll try to be very precise in the following, to hopefully clear up the remaining small pieces/thoughts.

> I went through your comments in the wrong order and missed that you already provided this solution for me! Oops.

I had the same thing with the docs of handleError, missed the fact that you already stated "Docs todo" in the commit message :)

> The purpose of result() is to synchronize the end-states of all parts of a stream. Now with #52, it also has the role of ensuring that errors passed to end() throw downstream, and errors thrown by end() throw upstream.

Well, the purpose of `result()` didn't really change. What did change is that its behavior was previously arguably a bit simpler for handling certain ender errors, but the user no longer had the option of actually responding to an error in an ender. Note that the only place where the change is really noticeable, is in a source, i.e. the very first element of a pipeline.

And for sources, the typical behavior is to not provide a value for result to `end()`, such that the default behavior kicks in, which is to 'bounce' any ender error that may occur.

> This mirrors the behaviour of write(). (...)
> The behaviour of end() is necessarily different, though. When the upstream gets an error by calling end(), it also can resolve it or possibly throw it upstream. What it can't do is "throw it back", because it's already called end() and cannot call it again.

Exactly. Otherwise, we could keep ping-ponging without end. It can "throw it back" through result, though. Either by passing the error back upstream, let it bounce there, and as such get forwarded downstream, or if it really wants to, it can influence the result by passing in e.g. a `defer()`'ed promise. The latter is probably an unlikely scenario in the real world.

> With #52, a stream can "throw it back" from end() by not including a result for end() to wait for. (...)

This was already the case, #52 didn't change anything for this. If you ended the stream without an error, and without a result, but the ender failed, it already 'bounced' that error into `result()`.

What changed is this:
- If you ended the stream *with* an error, and *without* a result, the final `result()` would be the error passed to `end()`, regardless of whether the ender threw. This has now been changed to become the ender's error if there was any, or otherwise the error passed to `end()`. I.e. in case you don't explicitly indicate that you manually want to handle any errors, you'll get sane results where any error that popped up will be reflected in the stream's result.
- If you ended the stream *with* an error, and *with* a result, the final `result()` would always be the error passed to `end()`, regardless of whether the ender threw. This has now been changed to always just become the result promise that was passed in, regardless of whether the ender threw (while of course still sending the ender rejection upstream to the originating `end()` call). This allows any upstream sources to decide if and how to handle any ender error. Again, what will typically happen is that it gets sent upstream all the way to a source, which calls `end()` without a result, and will default to bounce that error into `result()`, thereby passing all the way downstream again. The big difference is that if the source (or any other element in-between) *does* decide to process/swallow that error, they have a way to let the final result be successful.

> It seems to me that there is a missing capacity here, which is: how to resolve an error from end() when we're already at the source? We can do this if it's a failed call to write().

There are two options for this:
- If you're a source and the default behavior of bouncing any error into result is what you want (which I expect to be the case for most sources), just don't provide a result to end and you're done.
- If you're a source, or some other midway-stream and want fine-grained control, you'll have to do more work, and most likely provide a deferred promise as result, and resolve/reject that however you want.

> Using Promise.resolve() works, but it's clearly a hack—I'm just "lying" to the ender, telling it that there's an infinitely forgiving source still farther upstream that will take care of everything!

You should never do this in 'normal' code. The only reason I provided it here, in your test, is because we want to test explicitly whether any ender error is *only* passed upstream, and `result()` is indeed *only* reflecting the upstream's result.
Which is why I also added that explicit comment in that `end(...)` code.

'Normal' code will just not provide a result (for typical source), or just provide `upstream.result()` (for typical pipeline element).

> One could also do something like: (...)

Exactly.

> This is more "honest" in terms of the idiom, but not very easy to follow. (Maybe we could have a library function that wraps this behaviour? `endSource(error: Error | undefined, preemptEnderError: (enderError: Error) => void?)`

Yes. But as said, this kind of control will likely be very rare. We have the *option* of doing this, but most people probably won't. I doubt whether it's worth 'hiding' this in a helper, it's probably clearer to just spell it out in that weird transform that needs it :)

I do see some value in providing a transform that allows you to process errors like this though, as you can plug it into any pipeline that you want. On the other hand, chances are that someone else then needs something slightly different, and they'd still just write their own. In any case, this should then be a separate PR :)

> Back to the behaviour of batch, specifically: my conclusion is that things are working as they should. In some ways it seems like the ender of batch should be handling errors from flush outside of ender, since they're really write()-related, not end()-related. But that is a mixing of concerns that we've already committed to. My concerns here are really about #52 and not about batch().

On batch(), I agree that the behavior is now good, and I'm not too worried about 'handling' errors in the ender. It's a fact of life that for a transform like this, there is some chance that you'll need to do some cleanup work when the stream is ended. Similar to a line-splitter transform: if there's some remaining characters in the buffer when the stream ends, do you just discard them at that stage, or just pass them along at that last minute? Our current implementation is the latter. If really needed, we could even make that optional, especially for cases where batches are supposed to be exactly the same size. Don't really think that's necessary right now though.

I hope your concerns about #52 are now put to rest :)

Reply:
> This was already the case, #52 didn't change anything for this.

Yes. I should have said, "*As before*, a stream can..." Only the second part is different.

> I do see some value in providing a transform that allows you to process errors like this though, as you can plug it into any pipeline that you want.

Of course! That would be a good fix. This can always be addressed by just putting a pipe step directly after the source that handles ender errors as desired. Good point.

> On the other hand, chances are that someone else then needs something slightly different, and they'd still just write their own.

That doesn't bother me so much, as long as there is some kind of reference implementation for them to start from, especially since this isn't the most obvious behaviour. I do agree that it is a corner case though.

> In any case, this should then be a separate PR :)

No argument from me there :)

> If really needed, we could even make that optional, especially for cases where batches are supposed to be exactly the same size.

I would be interested to see if that case comes up. It's pretty trivial to handle downstream of course.

> I hope your concerns about #52 are now put to rest :)

They are, thank you :)

Reply:
(And thank you for taking the time to dig into and explain all the subtleties of this. I'm learning a lot! Not just about the particular decisions of `ts-stream` but also about the problem space.)